### PR TITLE
Various easy perf wins

### PR DESCRIPTION
**Pull request by chrisforbes**

### CODE REVIEW

**Comment by dneto0** [comment]
Rebased and pushed into master, ending in d431b69c28

File: source/validate.cpp [hunk 1]
```diff
@@ -326,7 +328,7 @@ spv_result_t ValidateBinaryUsingContextAndValidationState(
     spv_instruction_t inst;
     spvInstructionCopy(&binary->code[index], static_cast<SpvOp>(opcode),
                        wordCount, endian, &inst);
-    instructions.push_back(inst);
+    instructions.emplace_back(std::move(inst));
```
Comment (atgoo) [inline source/validate.cpp]: std::move decreases performance.

These two should work the same:
instructions.push_back(inst);
instructions.emplace_back(inst);

but
instructions.emplace_back(std::move(inst));
has an additional memory write to inst.

File: source/binary.cpp [hunk 2]
```diff
@@ -269,17 +273,18 @@ spv_result_t Parser::parseInstruction() {
...
+    _.endian_converted_words.reserve(25);
   }
 
   // After a successful parse of the instruction, the inst.operands member
   // will point to this vector's storage.
-  std::vector<spv_parsed_operand_t> operands;
   // Most instructions have fewer than 25 logical operands.
-  operands.reserve(25);
+  _.operands.clear();
+  _.operands.reserve(25);
```
Comment (atgoo) [inline source/binary.cpp]: This reserve can be done in the constructor.

File: source/binary.cpp [hunk 3]
```diff
@@ -269,17 +273,18 @@ spv_result_t Parser::parseInstruction() {
   // If the module's endianness is different from the host native endianness,
   // then converted_words contains the the endian-translated words in the
   // instruction.
-  std::vector<uint32_t> endian_converted_words = {first_word};
+  _.endian_converted_words.clear();
+  _.endian_converted_words.push_back(first_word);
   if (_.requires_endian_conversion) {
     // Most instructions have fewer than 25 words.
-    endian_converted_words.reserve(25);
+    _.endian_converted_words.reserve(25);
```
Comment (atgoo) [inline source/binary.cpp]: This reserve can be done in the constructor.

File: source/validate.cpp (hunk 1 above)
Comment (ratchetfreak) [inline source/validate.cpp]: but the copy (without std::move) needs an extra allocation which is then followed by a deallocation in the destructor of the original.

Comment (chrisforbes) [inline source/validate.cpp]: @atgoo, the issue here is the `std::vector` embedded in `inst`. I will take another look at it and undo this change if it's really losing.

File: source/binary.cpp (hunk 3 above)
Comment (chrisforbes) [inline source/binary.cpp]: Indeed it can. I will move those there, in both cases.

File: source/validate.cpp (hunk 1 above)
Comment (atgoo) [inline source/validate.cpp]: > the issue here is the std::vector embedded in inst. I will take another look at it and undo this change if it's really losing

Yes, if instruction is not POD, then what I said doesn't apply.

Comment (atgoo) [inline source/validate.cpp]: I see, I confused it with spv_parsed_instruction_t which is POD.

**Pull request closed without merging**